sockeye/lexical_constraints.py

```diff
@@ -24,106 +24,139 @@
 # Represents a list of raw constraints for a sentence. Each constraint is a list of target-word IDs.
 RawConstraintList = List[List[int]]
 
+class AvoidTrie:
+    """
+    Represents a set of phrasal constraints for an input sentence.
+    These are organized into a trie.
+    """
+    def __init__(self, raw_phrases: Optional[RawConstraintList] = None) -> None:
+        self.final_ids = set()  # type: Set[int]
+        self.children = {}      # type: Dict[int,'AvoidTrie']
+
+        if raw_phrases:
+            for phrase in raw_phrases:
+                self.add_phrase(phrase)
+
+    def __str__(self):
+        s = '({}'.format(list(self.final_ids))
+        for child_id in self.children.keys():
+            s += ' -> {} {}'.format(child_id, self.children[child_id])
+        s += ')'
+        return s
+
+    def add_phrase(self,
+                   phrase: List[int]) -> None:
+        if len(phrase) == 1:
+            self.final_ids.add(phrase[0])
+        else:
+            next_word = phrase[0]
+            if not next_word in self.children:
+                self.children[next_word] = AvoidTrie()
+            self.children[next_word].add_phrase(phrase[1:])
+
+    def step(self, word_id: int) -> 'AvoidTrie':
+        return self.children.get(word_id, None)
+
+    def final(self) -> Set[int]:
+        return self.final_ids
 
-class AvoidPhrase:
+
+class AvoidState:
```

Comment:
Do you need a separate class? Can't you just store AvoidTrie (perhaps renamed as AvoidTrieNode for clarity) in AvoidBatch?

Reply:
The state object tracks a position in the trie for the current hypothesis. We could just keep this position, but there are a number of associated operations that make sense to bundle together in a class.

Reply:
Yes, the question is if you can't bundle these operations in the trie and avoid yet another wrapper class. As you are already using a non-general, partly non-standard trie I think this would help clarity. But up to you, it's not a big issue.

Reply:
It would be simpler to have one fewer class. I'll give it some thought this evening.